File: midterm2/q02/Fibonacci.rb
module Fibonacci
  def self.fib(memo, position)
    return [1, memo] if position == 0 || position == 1
    return [memo[position], memo] if memo.key?(position)

    cur, memo_cur = fib(memo, position - 1)
    prev, memo_prev = fib(memo_cur, position - 2)
    result = cur + prev

    [result, memo_prev.merge(position => result).freeze]
  end

  def self.calculate_fibonacci(position)
    result, = fib({}.freeze, position)
    result
  end
end

position = 400
result = Fibonacci.calculate_fibonacci(position)
puts "Fibonacci number at position #{position} is: #{result}"

File: midterm2/q07/Fibonacci.fs
module Fibonacci =
    let rec fib (memo: Map<int, bigint>) (position: int) : bigint * Map<int, bigint> =
        if position = 0 || position = 1 then
            1I, memo
        elif memo.ContainsKey(position) then
            memo.[position], memo
        else
            let cur, memoCur = fib memo (position - 1)
            let prev, memoPrev = fib memoCur (position - 2)
            let result = cur + prev 

            result, memoPrev.Add(position, result)

    let calculateFibonacci position =
        let result, _ = fib Map.empty position
        result

[<EntryPoint>]
let main argv =
    let position = 400 
    let result = Fibonacci.calculateFibonacci position
    printfn "Fibonacci number at position %d is %A" position result
    0
